app/models/manageiq/providers/cisco_intersight/inventory/collector/physical_infra_manager.rb

```diff
@@ -3,133 +3,147 @@ class Inventory::Collector::PhysicalInfraManager < ManageIQ::Providers::Inventor
+      firmware_firmware_summaries.find { |c| c.server.moid == server_moid}
     end
 
-    def get_device_contract_informations
-      # Returns an array with objects of type DeviceContractInformation
-      get_asset_api.get_asset_device_contract_information_list.results
+    def get_source_object_from_physical_server(physical_summary)
+      # physical_summary represents API object, class IntersightClient::ComputePhysicalSummary
+      # Returns API object of either class IntersightClient::ComputeBlade or IntersightClient::ComputeRackUnit,
+      # depending on attribute source_object_type
+      source_object_type = physical_summary.source_object_type
+      source_object_moid = physical_summary.moid
+      if source_object_type == "compute.Blade"
+        get_compute_blade_by_moid(source_object_moid)
+      else
+        get_compute_rack_unit_by_moid(source_object_moid)
+      end
     end
 
-    def get_device_contract_information_from_device_moid(registered_device_moid)
-      get_device_contract_informations.select{|c| c.registered_device.moid == registered_device_moid}[0]
-    end
+    delegate :get_ether_physical_port_by_moid, :to => :ether_api
 
-    def get_equipment_locator_led_by_moid(moid)
-      get_equipment_api.get_equipment_locator_led_by_moid(moid)
-    end
+    delegate :get_port_group_by_moid, :to => :port_api
 
-    def get_storage_controller_by_moid(moid)
-      get_storage_api.get_storage_controller_by_moid(moid)
-    end
+    delegate :get_equipment_switch_card_by_moid, :to => :equipment_api
 
-    def get_compute_board_by_moid(moid)
-      get_compute_api.get_compute_board_by_moid(moid)
-    end
+    delegate :get_equipment_locator_led_by_moid, :to => :equipment_api
 
-    def get_adapter_unit_by_moid(moid)
-      get_adapter_api.get_adapter_unit_by_moid(moid)
-    end
+    delegate :get_storage_controller_by_moid, :to => :storage_api
 
-    def get_asset_device_registration_by_moid(moid)
-      get_asset_api.get_asset_device_registration_by_moid(moid)
-    end
+    delegate :get_compute_board_by_moid, :to => :compute_api
 
-    def get_rack_unit_from_physical_summary_moid(moid)
-      physical_racks.select { |c| c.registered_device.moid == moid }[0]
-    end
+    delegate :get_adapter_unit_by_moid, :to => :adapter_api
+
+    delegate :get_asset_device_registration_by_moid, :to => :asset_api
+
+    delegate :get_firmware_running_firmware_by_moid, :to => :firmware_api
+
+    delegate :get_compute_blade_by_moid, :to => :compute_api
 
-    def get_management_controller_by_moid(moid)
-      get_management_api.get_management_controller_by_moid(moid)
+    delegate :get_compute_rack_unit_by_moid, :to => :compute_api
+
+    private
+
+    # API endpoint declaration
+    def firmware_api
+      @firmware_api = IntersightClient::FirmwareApi.new
     end
 
-    def get_compute_blade_by_moid(moid)
-      get_compute_api.get_compute_blade_by_moid(moid)
+    def compute_api
+      @compute_api = IntersightClient::ComputeApi.new
```

Comment:
Hey @tjazsch so the idea with saving this to an instance variable is to prevent it from being initialized multiple times when it can be reused.
This is done with `@compute_api ||= IntersightClient::ComputeApi.new` that way the object is only initialized on first access.
You can pre-initialize all of these like you're doing now, but if they don't end up getting called then they're "wasted" objects.
Obviously for a full-refresh these will all be used but when we get to targeted refresh it is likely that you'll only need one or two of these.
So if you go the `||=` route then you don't need to initialize all of the api objects up [[here]](https://github.com/ManageIQ/manageiq-providers-cisco_intersight/pull/35/files#diff-7b43a5aefb4b0baee7620bb47b83cc7d4a040106c11f6de14c06390e9fc73c05R10-R19) and they will be initialized as the various methods need them

Reply:
Hey,

Thanks for taking your time and commenting. Will resolve this.